src/restic_fmt_plug.c

```diff
@@ -0,0 +1,334 @@
+ * modification, are permitted.
+ */
+
+#if FMT_EXTERNS_H
+extern struct fmt_main fmt_restic;
+#elif FMT_REGISTERS_H
+john_register_one(&fmt_restic);
+#endif
+
+#include <string.h>
+
+#ifdef _OPENMP
+#include <omp.h>
+#endif
+
+#include "aes.h"
+#include "arch.h"
+#include "base64_convert.h"
+#include "common.h"
+#include "formats.h"
+#include "jumbo.h"
+#include "misc.h"
+#include "options.h"
+#include "params.h"
+#include "poly1305-donna/poly1305-donna.h"
+#include "yescrypt/yescrypt.h"
+
+#define FORMAT_NAME "Restic Repository"
+#define FORMAT_LABEL "restic"
+#define FORMAT_TAG "$restic$"
+#define TAG_LENGTH sizeof(FORMAT_TAG) - 1
+#define ALGORITHM_NAME "PBKDF2/scrypt Poly1305"
+#define PLAINTEXT_LENGTH 125
+#define BINARY_SIZE 16 /* MAC */
+#define BINARY_ALIGN 4
+#define DATA_SIZE 160
+#define SALT_ALIGN 4
+#define MAC_SIZE 16
+#define NONCE_SIZE 16
+#define BENCHMARK_COMMENT ""
+#define BENCHMARK_LENGTH 7
+#define MIN_KEYS_PER_CRYPT 1
+#define MAX_KEYS_PER_CRYPT 1
+
+static int max_threads;
+static yescrypt_local_t *local;
+static char (*saved_key)[PLAINTEXT_LENGTH + 1];
+static uint32_t (*crypt_out)[BINARY_SIZE / sizeof(uint32_t)];
+
+typedef struct restic_salt {
+	int N;
+	int r;
+	int p;
+	char salt[64];
+	unsigned char data[DATA_SIZE];
+} restic_salt;
+
+static restic_salt *cur_salt;
+
+static void init(struct fmt_main *self)
+{
+	max_threads = 1;
+	local = mem_alloc(sizeof(*local) * max_threads);
+	int i;
+	for (i = 0; i < max_threads; i++)
+		yescrypt_init_local(&local[i]);
+
+	saved_key = mem_calloc(self->params.max_keys_per_crypt, sizeof(*saved_key));
+	crypt_out = mem_calloc(self->params.max_keys_per_crypt, sizeof(*crypt_out));
+}
+
+static void done(void)
+{
+	int i;
+	for (i = 0; i < max_threads; i++)
+		yescrypt_free_local(&local[i]);
+	MEM_FREE(local);
+	MEM_FREE(saved_key);
+	MEM_FREE(crypt_out);
+}
+
+static int valid(char *ciphertext, struct fmt_main *self)
+{
+	char *ctcopy, *keeptr, *p;
+
+	if (strncmp(ciphertext, FORMAT_TAG, TAG_LENGTH) != 0)
+		return 0;
+
+	ctcopy = strdup(ciphertext);
+	keeptr = ctcopy;
+	ctcopy += TAG_LENGTH;
+	if (((p = strtokm(ctcopy, "*")) == NULL) || (strcmp("scrypt", p) != 0)) /* KDF */
+		goto err;
+	if (((p = strtokm(NULL, "*")) == NULL) || !isdec(p)) /* scrypt: N */
+		goto err;
+	if (((p = strtokm(NULL, "*")) == NULL) || !isdec(p)) /* scrypt: r */
+		goto err;
+	if (((p = strtokm(NULL, "*")) == NULL) || !isdec(p)) /* scrypt: p */
+		goto err;
+	if (((p = strtokm(NULL, "*")) == NULL) || !ishexn(p, 64)) /* scrypt: salt */
+		goto err;
+	if (((p = strtokm(NULL, "*")) == NULL) || !ishexn(p, 160)) /* restic: data */
+		goto err;
+
+	MEM_FREE(keeptr);
+	return 1;
+err:
+	MEM_FREE(keeptr);
+	return 0;
+}
+
+void *restic_get_binary(char *ciphertext)
+{
+	static unsigned char c[BINARY_SIZE];
+	char *p;
+	int i;
+	p = strrchr(ciphertext, '*') + 1;
+	// Restic design doc: computed MAC will not match the last 16 bytes of the data
+	p += (160 - 16) * 2;
```

Comment:
If we skip most of the data here, then maybe we shouldn't have that skipped data in the "hashes" in the first place (revise the Python script to skip it)?

Also, I don't see how what the code does agrees with the comment. Doesn't the code skip most of data leaving only the last 16 bytes, and trying to match that? Yet the comment says "will not match the last 16 bytes". Confusing, and probably wrong.

Reply:
> If we skip most of the data here, then maybe we shouldn't have that
> skipped data in the "hashes" in the first place (revise the Python script to skip it)?

All 160 bytes from data are used:

- First 16 bytes: Poly1305-AES Nonce
- Bytes 16-144: The encrypted Master Key we calculate the MAC on
- Last 16 bytes: The MAC we need to match

I decided to leave the DATA in this untyped format because this is
also the way it is stored in the `restic`-JSON file and `restic` might add
additional KDFs/MACs in the future and use DATA differently.


>  Also, I don't see how what the code does agrees
> with the comment. Doesn't the code skip most of data leaving only the
> last 16 bytes, and trying to match that? Yet the comment says "will
> not match the last 16 bytes". Confusing, and probably wrong.

Sorry, I copied only a fraction of the design document (without the
context: "If the password is incorrect or the key file has been
tampered with").